src/compiler/sys.ts

```diff
@@ -297,6 +304,131 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const dirWatchers = createFileMap<DirWatcher>();
+                const recursiveDirWatchers = createFileMap<DirWatcher>();
+                const fileWatcherCallbacks = createFileMap<FileWatcherCallback>();
+                const dirWatcherCallbacks = createFileMap<DirWatcherCallback>();
+
+                const currentDirectory = process.cwd();
+                return { addFile, removeFile, addDir };
+
+                function addDir(dirName: string, callback: DirWatcherCallback, recursive?: boolean) {
+                    const dirPath = toPath(dirName, currentDirectory, getCanonicalPath);
+                    dirWatcherCallbacks.set(dirPath, callback);
+                    const { watcher, isRecursive } = addDirWatcher(dirPath, recursive);
+                    return {
+                        close: () => reduceDirWatcherRefCount(watcher, dirPath, isRecursive)
+                    };
+                }
+
+                function reduceDirWatcherRefCount(watcher: DirWatcher, dirPath: Path, isRecursive: boolean) {
+                    watcher.referenceCount -= 1;
+                    if (watcher.referenceCount <= 0) {
+                        watcher.close();
+                        if (isRecursive) {
+                            recursiveDirWatchers.remove(dirPath);
+                        }
+                        else {
+                            dirWatchers.remove(dirPath);
+                        }
+                    }
+                }
+
+                function addDirWatcher(dirPath: Path, recursive?: boolean): { watcher: DirWatcher, isRecursive: boolean } {
+                    let watchers: FileMap<DirWatcher>;
+                    const options: { persistent: boolean, recursive?: boolean } = { persistent: true };
+
+                    // Node 4.0 `fs.watch` function supports the "recursive" option on both OSX and Windows
+                    // (ref: https://github.com/nodejs/node/pull/2649 and https://github.com/Microsoft/TypeScript/issues/4643)
+                    if (isNode4OrLater() && recursive === true) {

```

Comment:
Yes it wouldn't work. The alternative is to add non-recursive watchers on every sub-folder recursively, which might explode to too many watchers if the recursive watcher is set to an unfortunate place.  Polling doesn't reflect changes in sub-folders as well, therefore can't help with this.

Or in this case should we just return an error message to warn the user?